### PR TITLE
Update 3D lights and shadows documentation for 4.0

### PR DESCRIPTION
**Pull request by Calinou**

### CODE REVIEW

File: tutorials/3d/lights_and_shadows.rst [hunk 1]
```diff
@@ -153,30 +184,58 @@ radius.
...
+If the objects being rendered are mostly irregular and subdivided, Dual
+Paraboloid is usually enough. In any case, as these shadows are cached in a
+shadow atlas (more on that at the end), it may not make a difference in
+performance for most scenes.
 
-Omni Shadows can be rendered as either **"Dual Paraboloid" or "Cube Mapped"**.
-The former renders quickly, but can cause deformations,
-while the later is more correct, but costlier.
+Omni lights with shadows enabled can make use of projectors. The projector
+texture will *multiply* the lights's color by the color at a given point on the
```
Comment (skyace65) [inline tutorials/3d/lights_and_shadows.rst:221]:
```suggestion
texture will *multiply* the light's color by the color at a given point on the
```

File: tutorials/3d/lights_and_shadows.rst [hunk 2]
```diff
@@ -186,34 +245,59 @@ Spot lights are similar to omni lights, except they emit light only into a cone
...
 
-Spots don't need any parameters for shadow mapping. Keep in mind that, at more than 89 degrees of aperture, shadows
-stop functioning for spots, and you should consider using an Omni light instead.
+Spots feature the same parameters as omni lights for shadow mapping. Rendering
+spot shadow maps is significantly faster compared to omni lights, as only one
+shadow texture needs to be rendered (instead of rendering 6 faces, or 2 in dual
+parabolid mode).
+
+Spot lights with shadows enabled can make use of projectors. The projector
+texture will *multiply* the lights's color by the color at a given point on the
```
Comment (skyace65) [inline tutorials/3d/lights_and_shadows.rst:263]:
```suggestion
texture will *multiply* the light's color by the color at a given point on the
```

File: tutorials/3d/lights_and_shadows.rst [hunk 3]
```diff
@@ -48,30 +50,54 @@ Lights can optionally cast shadows. This gives them greater realism (light does
...
+- **Enabled:** Check to enable shadow mapping in this light.
+- **Opacity:** Areas occluded are darkened by this opacity factor. Shadows are
+  fully opaque by default, but this can be changed to make shadows translucent
+  for a given light.
+- **Bias:** When this parameter is too small, self shadowing occurs. When too
+  large, shadows separate from the casters. Tweak to what works best for you.
+- **Contact:** Performs a short screen-space raycast to reduce the gap generated
+  by the bias. Contact shadows are only available when using the GLES3 backend.
+- **Reverse Cull Face:** Some scenes work better when shadow mapping is rendered
+  with face-culling inverted.
```
Comment (skyace65) [inline tutorials/3d/lights_and_shadows.rst:63]: Several things about the above section.
- Contact is no longer a parameter at least in beta 4
- Transmittance bias, normal bias and blur are all missing from this list.

**Review by skyace65 — CHANGES_REQUESTED** [review]
Overall this looks very good, just needs a few changes.

**Review by skyace65 — APPROVED** [review]
Looks good now!

**Review by mhilbrunner — APPROVED** [review]

File: tutorials/3d/lights_and_shadows.rst [hunk 4]
```diff
@@ -9,74 +9,153 @@ Introduction
...
+real-time lighting. As many lights as desired can be added (as long as
+performance allows). However, there's still a default limit of 512 *clustered
+elements* that can be present in the current camera view. A clustered element is
+an omni light, a spot light, a :ref:`decal <doc_decals>` or a :ref:`reflection
+probe <doc_reflection_probes>`. This limit can be increased by adjusting the
+**Rendering > Limits > Cluster Builder > Max Clustered Elements** advanced
+project setting.
+
+When using the Forward Mobile renderer, there is a limitation of 8 OmniLights +
+8 SpotLights per mesh resource. This currently cannot be changed.
```
Comment (clayjohn) [inline tutorials/3d/lights_and_shadows.rst:64]: There is also a limit of 256 lights per scene, also not configurable

File: tutorials/3d/lights_and_shadows.rst [hunk 5]
```diff
@@ -9,74 +9,153 @@ Introduction
...
+
+When using the Compatibility renderer, up to 8 OmniLights + 8 SpotLights can be
+rendered per mesh resource. This limit can be increased in the advanced Project
+Settings by adjusting **Rendering > Limits > OpenGL > Max Renderable Lights**
+and/or **Rendering > Limits > OpenGL > Max Lights Per Object** at the cost of
+performance and longer shader compilation times. The limit can also be decreased
+to reduce shader compilation times and improve performance slightly.
+Additionally, only one DirectionalLight can be visible at a time.
```
Comment (clayjohn) [inline tutorials/3d/lights_and_shadows.rst:72]: Multiple DirectionalLight3Ds can be visible at the same time. You can just remove this whole line

File: tutorials/3d/lights_and_shadows.rst [hunk 6]
```diff
@@ -231,20 +373,142 @@ Every frame, the following procedure is performed for all lights:
...
+make use of the increased sample count.
+
+16-bits versus 24-bit
+^^^^^^^^^^^^^^^^^^^^^
+
+By default, Godot uses 16-bit depth textures for shadow map rendering. This is
+recommended in most cases as it performs better without a noticeable difference
+in quality.
+
+If **16 Bits** is disabled, 24-bit shadow rendering will be used instead. This
```
Comment (clayjohn) [inline tutorials/3d/lights_and_shadows.rst:452]:
```suggestion
If **16 Bits** is disabled, 32-bit depth textures will be used instead. This
```

**Review by clayjohn — COMMENTED** [review]
Amazing work! Just a few comments. This is nearly ready to merge

File: tutorials/3d/lights_and_shadows.rst (hunk 6 above)
Comment (Calinou) [reply tutorials/3d/lights_and_shadows.rst:452]: I thought we were using 24-bit depth instead, but I guess it depends on hardware support (we should be using 32-bit everywhere nowadays, anyway).

File: tutorials/3d/lights_and_shadows.rst (hunk 4 above)
Comment (Calinou) [reply tutorials/3d/lights_and_shadows.rst:64]: Is this 256 OmniLights + 256 SpotLights, or 256 OmniLights/SpotLights (meaning you can't have more than 128 of each if you want a 50-50 split)?

File: tutorials/3d/lights_and_shadows.rst (hunk 5 above)
Comment (Calinou) [reply tutorials/3d/lights_and_shadows.rst:72]: This is a leftover I forgot to remove – I can confirm multiple lights can be seen when using OpenGL.

File: tutorials/3d/lights_and_shadows.rst (hunk 4 above)
Comment (clayjohn) [reply tutorials/3d/lights_and_shadows.rst]: You can have 256 Omnilights and 256 Spotlights (so up to a total of 512)

**Review by clayjohn — APPROVED** [review]
Great work! Looks good to me

**Comment by akien-mga** [comment]
Can be rebased to fix warnings now that the page on decals was merged.

**Pull request merged**